### ISSUE TITLE
BUG: DeepExplainer (PyTorch) leaks memory due to traversal bug in remove_attributes

### ISSUE THREAD

**Issue by AftAb-25** [shap/shap#4839]
## Description

I've been dealing with a growing memory leak and what looks like stale tensor data when running `DeepExplainer` on some of my PyTorch models. After digging into the internals, I realized it's happening because of how [remove_attributes](cci:1://file:///c:/Users/aftab/Desktop/shap/shap/explainers/_deep/deep_pytorch.py:79:4-94:24) traverses the model compared to [add_handles](cci:1://file:///c:/Users/aftab/Desktop/shap/shap/explainers/_deep/deep_pytorch.py:65:4-77:27).

When [add_handles](cci:1://file:///c:/Users/aftab/Desktop/shap/shap/explainers/_deep/deep_pytorch.py:65:4-77:27) registers the forward/backward hooks, it correctly recurses through any module that has children (using `list(model.children())`). It only registers hooks once it hits the leaf modules.

However, [remove_attributes](cci:1://file:///c:/Users/aftab/Desktop/shap/shap/explainers/_deep/deep_pytorch.py:79:4-94:24) (around line 80 in [deep_pytorch.py](cci:7://file:///c:/Users/aftab/Desktop/shap/shap/explainers/_deep/deep_pytorch.py:0:0-0:0)) tries to decide if it should recurse by checking `"nn.modules.container" in str(type(child))`. Because of this brittle string check, it only recurses into things like `Sequential` or `ModuleList`. If you have a custom wrapper like a ResNet block, an attention layer, or just a generic custom `nn.Module` subclass, the string check fails. 

Instead of recursing into the custom block to clean up the `.x` and `.y` tensors on the leaf layers inside, [remove_attributes](cci:1://file:///c:/Users/aftab/Desktop/shap/shap/explainers/_deep/deep_pytorch.py:79:4-94:24) treats the custom block as a leaf itself. It silently tries to delete `.x` and `.y` on the container, which does nothing, leaving the stale tensors permanently attached to the nested leaf layers. Because these tensors retain their computation graph and memory, we get a massive leak, and the hooks can reuse stale values on subsequent [shap_values](cci:1://file:///c:/Users/aftab/Desktop/shap/shap/explainers/_tree.py:505:4-716:49) calls!

## Expected Behavior
Calling `explainer.shap_values(X)` should clean up all the `.x` and `.y` attribute tensors it temporarily adds to the model's leaf nodes during the forward hook process. Subsequent calls to [shap_values(X)](cci:1://file:///c:/Users/aftab/Desktop/shap/shap/explainers/_tree.py:505:4-716:49) should not have interference from previous intermediate hook outputs.

## Actual Behavior
The temporary `.x` and `.y` tensors are permanently left behind on any leaf module that sits inside a custom `nn.Module` wrapper class. Because these tensors retain their autograd history and gradients attached to the graph, memory usage balloons rapidly when repeatedly calling [shap_values](cci:1://file:///c:/Users/aftab/Desktop/shap/shap/explainers/_tree.py:505:4-716:49) in a loop.

## To Reproduce

Here's a minimal example showing the bug in action. I created a custom block that isn't under `nn.modules.container`.

```python
import torch
import torch.nn as nn
import numpy as np
import shap

# A custom block that will fail the `nn.modules.container` check
class CustomBlock(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(5, 5)
        self.act = nn.ReLU()
        
    def forward(self, x):
        return self.act(self.fc(x))

class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.block = CustomBlock()
        self.out = nn.Linear(5, 1)
        
    def forward(self, x):
        return self.out(self.block(x))

model = MyModel()
background = torch.randn(50, 5)
explainer = shap.DeepExplainer(model, background)
sv = explainer.shap_values(torch.randn(3, 5))

# After shap_values(), these leaf layers inside CustomBlock SHOULD be cleaned up.
# But instead, the tensors are permanently leaked:
print("fc has .x:", hasattr(model.block.fc, 'x'))    # Output: True (LEAKED!)
print("act has .x:", hasattr(model.block.act, 'x'))  # Output: True (LEAKED!)
```
## Proposed Fix
We can easily fix this by removing the str(type(child)) check entirely inside remove_attributes. We just need to make the tree traversal logic identical to the one currently used by add_handles—basically, if list(model.children()) is populated, recurse into them. Otherwise, run the cleanup block for the leaf module.

## Installed Version
SHAP version: Latest master branch (0.47.0.dev0) / Also present in current stable releases
PyTorch version: 2.x+ (Any PyTorch version)

### Bug report checklist

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [x] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [x] I'd be interested in making a PR to fix this bug
